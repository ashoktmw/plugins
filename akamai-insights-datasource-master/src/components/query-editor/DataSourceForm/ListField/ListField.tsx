import {
  IconButton,
  InlineField,
  InlineFieldRow
} from '@grafana/ui';
import React, { Fragment, JSX } from 'react';

import { ListIconsRow } from '../../../common/styled/ListIconsRow';
import { shortLabelWidth } from '../FormTypes';

export interface ListFieldProps<T> {
  fieldLabel: string;
  fieldTooltip?: any;
  resource: string;
  rows: T[];
  modelProvider: () => T;
  onChange: (list: T[]) => void;
  Editor: (model: T, index: number, onChange: (updatedModel: T, index: number) => void) => JSX.Element;
}

export function ListField<T>({ fieldLabel, fieldTooltip, resource, rows, modelProvider, onChange, Editor }: ListFieldProps<T>): JSX.Element {
  const onAdd = () => onChange([ ...rows, modelProvider() ]);

  const onRemoveAll = () => onChange([]);

  const onRemove = (index: number) => {
    const listCopy = [ ...rows ];
    listCopy.splice(index, 1);

    onChange(listCopy);
  };

  const onChangeSingle = (updatedModel: T, index: number) => {
    onChange(rows.map((model, i) => {
      if (i === index) {
        return {
          ...model,
          ...updatedModel
        };
      }

      return model;
    }));
  };
  
  return (
    <Fragment>
      <InlineFieldRow>
        <InlineField
          label={fieldLabel}
          tooltip={fieldTooltip}
          labelWidth={shortLabelWidth}>
          <ListIconsRow>
            {rows?.length ? <IconButton name="trash-alt" tooltip="Remove all" onClick={onRemoveAll}/> : ''}
            <IconButton
              name="plus-circle"
              tooltip={`Add new ${resource}`}
              onClick={onAdd}
            />
          </ListIconsRow>
        </InlineField>
      </InlineFieldRow>

      {rows.map((model, index) => (
        <InlineFieldRow key={index}>
          { Editor(model, index, updated => onChangeSingle(updated, index)) }

          <InlineField>
            <ListIconsRow>
              <IconButton
                name="minus-circle"
                tooltip={`Remove ${resource}`}
                onClick={() => onRemove(index)}
              />
            </ListIconsRow>
          </InlineField>
        </InlineFieldRow>)
      )}
    </Fragment>
  );
}
