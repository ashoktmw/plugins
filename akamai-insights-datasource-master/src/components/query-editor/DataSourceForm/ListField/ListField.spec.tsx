import { Input } from '@grafana/ui';
import React from 'react';
import renderer from 'react-test-renderer';

import { ListField } from './ListField';

describe('ListField component', () => {
  it('should render list with custom row Editor', () => {
    const rows: string[] = [ 'value 1', 'value 2' ] ;
    const onChange = jest.fn();

    const tree = renderer.create(
      <ListField
        fieldLabel="List label"
        fieldTooltip="List tooltip"
        resource="resource"
        modelProvider={() => ''}
        rows={rows}
        Editor={(value, index) => <Input key={index} value={value}/>}
        onChange={onChange}/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
