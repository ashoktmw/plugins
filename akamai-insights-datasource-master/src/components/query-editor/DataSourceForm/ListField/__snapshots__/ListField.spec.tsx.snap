// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListField component should render list with custom row Editor 1`] = `
[
  <div
    className="css-9doeo5-InlineFieldRow"
  >
    <div
      className="css-1f5gueu"
    >
      <label
        className="css-1e5y6i9"
      >
        List label
        <div
          className="css-wf08df-Icon"
          tabIndex={0}
        >
          <svg
            data-testid="info-circle"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          />
        </div>
      </label>
      <div
        className="css-1rfunm5"
      >
        <div
          className="css-1uedr6a"
        >
          <button
            aria-label="Remove all"
            className="css-x1vujn"
            onClick={[Function]}
            tabIndex={0}
          >
            <div
              className="css-wf08df-Icon"
            >
              <svg
                data-testid="trash-alt"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </button>
          <button
            aria-label="Add new resource"
            className="css-x1vujn"
            onClick={[Function]}
            tabIndex={0}
          >
            <div
              className="css-wf08df-Icon"
            >
              <svg
                data-testid="plus-circle"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="css-9doeo5-InlineFieldRow"
  >
    <div
      className="css-xcstkt-input-wrapper"
      data-testid="input-wrapper"
    >
      <div
        className="css-1w5c5dq-input-inputWrapper"
      >
        <input
          className="css-1n6y0bv-input-input"
          style={
            {
              "paddingLeft": undefined,
              "paddingRight": undefined,
            }
          }
          value="value 1"
        />
      </div>
    </div>
    <div
      className="css-1f5gueu"
    >
      <div
        className="css-1rfunm5"
      >
        <div
          className="css-1uedr6a"
        >
          <button
            aria-label="Remove resource"
            className="css-x1vujn"
            onClick={[Function]}
            tabIndex={0}
          >
            <div
              className="css-wf08df-Icon"
            >
              <svg
                data-testid="minus-circle"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="css-9doeo5-InlineFieldRow"
  >
    <div
      className="css-xcstkt-input-wrapper"
      data-testid="input-wrapper"
    >
      <div
        className="css-1w5c5dq-input-inputWrapper"
      >
        <input
          className="css-1n6y0bv-input-input"
          style={
            {
              "paddingLeft": undefined,
              "paddingRight": undefined,
            }
          }
          value="value 2"
        />
      </div>
    </div>
    <div
      className="css-1f5gueu"
    >
      <div
        className="css-1rfunm5"
      >
        <div
          className="css-1uedr6a"
        >
          <button
            aria-label="Remove resource"
            className="css-x1vujn"
            onClick={[Function]}
            tabIndex={0}
          >
            <div
              className="css-wf08df-Icon"
            >
              <svg
                data-testid="minus-circle"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;
